### PR TITLE
Compression Streams Handle Out of Memory Conditions

### PR DESCRIPTION
**Pull request by stwrt**
#### 7a01e745839ab2ac19fe60d0b619034a2339a6cf
<pre>
Compression Streams Handle Out of Memory Conditions
<a href="https://bugs.webkit.org/show_bug.cgi?id=244411">https://bugs.webkit.org/show_bug.cgi?id=244411</a>

Reviewed by Youenn Fablet.

Compression Streams scales up memory quite quickly, and can exhaust memory on
a lower spec device. This patch will scale down memory usage if allocations start failing.

A guard limit of 16KB of free memory was added. If free memory falls below that limit we will give up and
report to the stream not enough memory is avaiable to continue.

* Source/WebCore/Modules/compression/CompressionStreamEncoder.cpp:
(WebCore::CompressionStreamEncoder::compress):
* Source/WebCore/Modules/compression/DecompressionStreamDecoder.cpp:
(WebCore::DecompressionStreamDecoder::decompressZlib):
(WebCore::DecompressionStreamDecoder::decompressAppleCompressionFramework):

Canonical link: <a href="https://commits.webkit.org/253930@main">https://commits.webkit.org/253930@main</a>
</pre>

https://github.com/WebKit/WebKit/commit/6081506706c297c83df820db2129ba927753cafe

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/87475 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/69/builds/31564 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/18276 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/96701 "Built successfully") | [✅ 🛠 🧪 win](https://ews-build.webkit.org/#/builders/10/builds/150151 "Built successfully and passed tests") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/91448 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/64/builds/29928 "Built successfully") | [✅ 🛠 mac-debug](https://ews-build.webkit.org/#/builders/71/builds/26107 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/79590 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/91470 "Built successfully") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/93094 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/68/builds/24178 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/74257 "Built successfully") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/24014 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/9/builds/79145 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/79365 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/34/builds/67032 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/66/builds/27637 "Built successfully") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/73/builds/13197 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/67/builds/27589 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/14213 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/2763 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/63/builds/29279 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/60/builds/37078 "Passed tests") | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/65/builds/29205 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/33491 "Passed tests") | |

### CODE REVIEW

File: Source/WebCore/Modules/compression/CompressionStreamEncoder.cpp [hunk 1]
```diff
@@ -114,7 +114,16 @@ ExceptionOr<RefPtr<JSC::ArrayBuffer>> CompressionStreamEncoder::compress(const u
     }
 
     while (shouldCompress) {
-        auto output = Vector<uint8_t>(allocateSize);
+        if (allocateSize < startingAllocationSize)
+            return Exception { TypeError, "Out of Memory."_s };
```
Comment (youennf) [inline Source/WebCore/Modules/compression/CompressionStreamEncoder.cpp]: Should probably be Exception { OutOfMemoryError }.
Ditto for decompressor.

File: Source/WebCore/Modules/compression/CompressionStreamEncoder.cpp [hunk 2]
```diff
@@ -114,7 +114,16 @@ ExceptionOr<RefPtr<JSC::ArrayBuffer>> CompressionStreamEncoder::compress(const u
...
+
+        auto output = Vector<uint8_t>();
+        if (!output.tryReserveInitialCapacity(allocateSize)) {
+            allocateSize /= 4;
+            continue;
+        }
+
+        output.resize(allocateSize);
```
Comment (youennf) [inline Source/WebCore/Modules/compression/CompressionStreamEncoder.cpp]: Why output.resize(allocateSize), it seems unnecessary here given tryReserveInitialCapacity succeeded
storage.takeAsArrayBuffer() may return nullptr if not enough memory.
We should probably handle that case as well.
Something like:
```
auto result = storage.takeAsArrayBuffer();
if (!result)
     return Exception { OutOfMemoryError };
```

File: Source/WebCore/Modules/compression/CompressionStreamEncoder.cpp [hunk 3]
```diff
@@ -114,7 +114,16 @@ ExceptionOr<RefPtr<JSC::ArrayBuffer>> CompressionStreamEncoder::compress(const u
...
 
     while (shouldCompress) {
-        auto output = Vector<uint8_t>(allocateSize);
+        if (allocateSize < startingAllocationSize)
+            return Exception { TypeError, "Out of Memory."_s };
+
+        auto output = Vector<uint8_t>();
+        if (!output.tryReserveInitialCapacity(allocateSize)) {
+            allocateSize /= 4;
+            continue;
```
Comment (youennf) [inline Source/WebCore/Modules/compression/CompressionStreamEncoder.cpp]: Why not doing the if (allocateSize < startingAllocationSize) within this if check?

File: Source/WebCore/Modules/compression/DecompressionStreamDecoder.cpp [hunk 4]
```diff
@@ -118,7 +118,16 @@ ExceptionOr<RefPtr<JSC::ArrayBuffer>> DecompressionStreamDecoder::decompressZlib
...
 
     while (shouldDecompress) {
-        auto output = Vector<uint8_t>(allocateSize);
+        if (allocateSize < startingAllocationSize)
+            return Exception { TypeError, "Out of Memory."_s };
+
+        auto output = Vector<uint8_t>();
+        if (!output.tryReserveInitialCapacity(allocateSize)) {
+            allocateSize /= 4;
+            continue;
```
Comment (youennf) [inline Source/WebCore/Modules/compression/DecompressionStreamDecoder.cpp]: Here allocateSize is always initially equal to startingAllocationSize.
If the first tryReserveInitialCapacity fails, we will always return the exception.

File: Source/WebCore/Modules/compression/DecompressionStreamDecoder.cpp [hunk 5]
```diff
@@ -176,7 +185,16 @@ ExceptionOr<RefPtr<JSC::ArrayBuffer>> DecompressionStreamDecoder::decompressAppl
...
     
     while (shouldDecompress) {
-        auto output = Vector<uint8_t>(allocateSize);
+        if (allocateSize < startingAllocationSize)
+            return Exception { TypeError, "Out of Memory."_s };
+
+        auto output = Vector<uint8_t>();
+        if (!output.tryReserveInitialCapacity(allocateSize)) {
+            allocateSize /= 4;
+            continue;
```
Comment (youennf) [inline Source/WebCore/Modules/compression/DecompressionStreamDecoder.cpp]: Ditto

File: Source/WebCore/Modules/compression/CompressionStreamEncoder.cpp (hunk 2 above)
Comment (stwrt) [reply Source/WebCore/Modules/compression/CompressionStreamEncoder.cpp]: The resize here should be just updating the internal m_size variable. This was needed to handle some behavior later with the code. I am not a huge fan of that line of code either. Will see if I can clean it up.

I fully agree on needing to check the result of `storage.takeAsArrayBuffer()`.

File: Source/WebCore/Modules/compression/DecompressionStreamDecoder.cpp (hunk 4 above)
Comment (stwrt) [reply Source/WebCore/Modules/compression/DecompressionStreamDecoder.cpp]: Exactly 😀.

We should have a minimum expectation of free memory that we can allocate for this operation. I think 16KB of memory is not unreasonable.

If you have a differing opinion on how to handle this please feel free to mention it.

File: Source/WebCore/Modules/compression/DecompressionStreamDecoder.cpp [hunk 6]
```diff
@@ -176,7 +190,16 @@ ExceptionOr<RefPtr<JSC::ArrayBuffer>> DecompressionStreamDecoder::decompressAppl
     m_stream.src_size = inputLength;
     
     while (shouldDecompress) {
-        auto output = Vector<uint8_t>(allocateSize);
+        if (allocateSize < startingAllocationSize)
+            return Exception { TypeError, "Out of Memory."_s };
```
Comment (youennf) [inline Source/WebCore/Modules/compression/DecompressionStreamDecoder.cpp]: Should be OutOfMemoryError probably

**Review by youennf — APPROVED** [review]

File: Source/WebCore/Modules/compression/DecompressionStreamDecoder.cpp (hunk 6 above)
Comment (stwrt) [reply Source/WebCore/Modules/compression/DecompressionStreamDecoder.cpp]: Oops missed that 😀

**Pull request closed without merging**